Optimize Logs SDK BatchProcessor

wmdanor opened the pull request:
Towards: https://github.com/open-telemetry/opentelemetry-go/issues/6382

This PR optimizes Logs SDK BatchProcessor.

## Why? <!-- was i so blind -->

Batch processor does not check if buffer exporter is full, instead it clones records to buffer and then tries to push them to buffer exporter, and then exporter will return false from EnqueueExport.

It does not cause any performance issues when using NOOP exporter, because it "exports" instantly. -> Buffer exporter input channel will almost never be full.

But actual implementations of exporters are not going to act that way, they need some time for export operation. And so buffer exporter will get full fairly frequently. And each export attempt to buffer exporter costs us copying of entire queue inside of batch processor.

In order to catch this or similar performance bottleneck in the future, new benchmark (`BatchSimulateExport`) was added which will use exporter that will simulate non-instant export by sleeping for 5ms.

## Benchmarks

```sh
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/log
cpu: AMD Ryzen 9 7900X 12-Core Processor
                                 │   base.txt   │               new.txt               │
                                 │    sec/op    │   sec/op     vs base                │
Processor/Simple-24                 340.9n ± 2%   343.5n ± 5%        ~ (p=0.971 n=10)
Processor/Batch-24                  667.6n ± 2%   666.3n ± 4%        ~ (p=0.896 n=10)
Processor/BatchSimulateExport-24   4441.5n ± 5%   526.4n ± 4%  -88.15% (p=0.000 n=10)
Processor/SetTimestampSimple-24     360.0n ± 4%   351.0n ± 1%        ~ (p=0.247 n=10)
Processor/SetTimestampBatch-24      669.8n ± 3%   647.0n ± 5%        ~ (p=0.052 n=10)
Processor/AddAttributesSimple-24    383.9n ± 1%   386.6n ± 5%        ~ (p=0.138 n=10)
Processor/AddAttributesBatch-24     759.9n ± 6%   734.5n ± 5%        ~ (p=0.481 n=10)
Processor/SetAttributesSimple-24    369.9n ± 2%   372.4n ± 3%        ~ (p=0.643 n=10)
Processor/SetAttributesBatch-24     684.2n ± 4%   674.0n ± 8%        ~ (p=0.529 n=10)
geomean                             639.9n        499.9n       -21.88%

                                 │   base.txt   │                new.txt                 │
                                 │     B/op     │     B/op      vs base                  │
Processor/Simple-24                  450.0 ± 0%     450.0 ± 0%        ~ (p=1.000 n=10) ¹
Processor/Batch-24                 1.153Ki ± 2%   1.125Ki ± 1%   -2.41% (p=0.001 n=10)
Processor/BatchSimulateExport-24     649.0 ± 1%     473.0 ± 0%  -27.12% (p=0.000 n=10)
Processor/SetTimestampSimple-24      450.0 ± 0%     450.0 ± 0%        ~ (p=1.000 n=10) ¹
Processor/SetTimestampBatch-24     1.145Ki ± 2%   1.122Ki ± 2%   -2.01% (p=0.003 n=10)
Processor/AddAttributesSimple-24     450.0 ± 0%     450.0 ± 0%        ~ (p=1.000 n=10) ¹
Processor/AddAttributesBatch-24    1.169Ki ± 2%   1.148Ki ± 1%   -1.80% (p=0.000 n=10)
Processor/SetAttributesSimple-24     450.0 ± 0%     450.0 ± 0%        ~ (p=1.000 n=10) ¹
Processor/SetAttributesBatch-24    1.141Ki ± 2%   1.128Ki ± 2%   -1.11% (p=0.030 n=10)
geomean                              719.3          688.7        -4.24%
¹ all samples are equal

                                 │  base.txt  │               new.txt               │
                                 │ allocs/op  │ allocs/op   vs base                 │
Processor/Simple-24                1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/Batch-24                 1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/BatchSimulateExport-24   1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetTimestampSimple-24    1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetTimestampBatch-24     1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/AddAttributesSimple-24   1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/AddAttributesBatch-24    1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetAttributesSimple-24   1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
Processor/SetAttributesBatch-24    1.000 ± 0%   1.000 ± 0%       ~ (p=1.000 n=10) ¹
geomean                            1.000        1.000       +0.00%
¹ all samples are equal
```

## Benchmarks from issue (external)

```sh
goos: linux
goarch: amd64
pkg: github.com/pellared/spanevents-vs-logs
cpu: AMD Ryzen 9 7900X 12-Core Processor
           │ spanevents.txt │              logs.base.txt              │             logs.new.txt             │
           │     sec/op     │    sec/op      vs base                  │   sec/op     vs base                 │
/OTLP-24        41.40µ ± 3%   2935.15µ ± 9%  +6989.05% (p=0.000 n=10)   84.19µ ± 4%  +103.34% (p=0.000 n=10)
/STDOUT-24      39.55µ ± 4%    623.45µ ± 1%  +1476.21% (p=0.000 n=10)   89.51µ ± 3%  +126.29% (p=0.000 n=10)
/NOOP-24       13.179µ ± 2%     4.950µ ± 1%    -62.44% (p=0.000 n=10)   4.932µ ± 1%   -62.58% (p=0.000 n=10)
geomean         27.84µ          208.4µ        +648.68%                  33.37µ        +19.86%

           │ spanevents.txt │             logs.base.txt              │             logs.new.txt             │
           │      B/op      │     B/op       vs base                 │     B/op      vs base                │
/OTLP-24       98.90Ki ± 8%   210.44Ki ± 0%  +112.79% (p=0.000 n=10)   52.01Ki ± 0%  -47.41% (p=0.000 n=10)
/STDOUT-24     91.01Ki ± 0%   286.62Ki ± 0%  +214.93% (p=0.000 n=10)   81.67Ki ± 1%  -10.27% (p=0.000 n=10)
/NOOP-24       29840.0 ± 0%      240.0 ± 0%   -99.20% (p=0.000 n=10)     240.0 ± 0%  -99.20% (p=0.000 n=10)
geomean        64.01Ki         24.18Ki        -62.23%                  9.985Ki       -84.40%

           │ spanevents.txt │            logs.base.txt             │            logs.new.txt            │
           │   allocs/op    │  allocs/op   vs base                 │ allocs/op   vs base                │
/OTLP-24        699.0 ± 20%   2174.5 ± 1%  +211.09% (p=0.000 n=10)   249.5 ± 1%  -64.31% (p=0.000 n=10)
/STDOUT-24      646.5 ±  1%   4912.5 ± 0%  +659.86% (p=0.000 n=10)   872.5 ± 2%  +34.96% (p=0.000 n=10)
/NOOP-24      303.000 ±  0%    3.000 ± 0%   -99.01% (p=0.000 n=10)   3.000 ± 0%  -99.01% (p=0.000 n=10)
geomean         515.4          317.6        -38.37%                  86.76       -83.17%
```

```sh
goos: linux
goarch: amd64
pkg: github.com/pellared/spanevents-vs-logs
cpu: AMD Ryzen 9 7900X 12-Core Processor
           │ logs.base.txt │            logs.new.txt             │
           │    sec/op     │   sec/op     vs base                │
/OTLP-24     2935.15µ ± 9%   84.19µ ± 4%  -97.13% (p=0.000 n=10)
/STDOUT-24    623.45µ ± 1%   89.51µ ± 3%  -85.64% (p=0.000 n=10)
/NOOP-24       4.950µ ± 1%   4.932µ ± 1%        ~ (p=0.342 n=10)
geomean        208.4µ        33.37µ       -83.99%

           │ logs.base.txt │              logs.new.txt              │
           │     B/op      │     B/op      vs base                  │
/OTLP-24     210.44Ki ± 0%   52.01Ki ± 0%  -75.28% (p=0.000 n=10)
/STDOUT-24   286.62Ki ± 0%   81.67Ki ± 1%  -71.51% (p=0.000 n=10)
/NOOP-24        240.0 ± 0%     240.0 ± 0%        ~ (p=1.000 n=10) ¹
geomean       24.18Ki        9.985Ki       -58.70%
¹ all samples are equal

           │ logs.base.txt │             logs.new.txt             │
           │   allocs/op   │ allocs/op   vs base                  │
/OTLP-24       2174.5 ± 1%   249.5 ± 1%  -88.53% (p=0.000 n=10)
/STDOUT-24     4912.5 ± 0%   872.5 ± 2%  -82.24% (p=0.000 n=10)
/NOOP-24        3.000 ± 0%   3.000 ± 0%        ~ (p=1.000 n=10) ¹
geomean         317.6        86.76       -72.69%
¹ all samples are equal
```

wmdanor commented on sdk/log/exporter.go:
```diff
@@ -189,8 +189,8 @@ type bufferExporter struct {
 func newBufferExporter(exporter Exporter, size int) *bufferExporter {
-	if size < 0 {
-		size = 0
+	if size < 1 {
```
This change is needed because unbuffered channel is always "full" and requires data to be pulled from other side.

pellared replied:
Makes sense to me. Maybe this can be even done in a separate PR as it seems a good change even without the other changes?

pellared commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,17 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
-				return ok
-			})
+			qLen := b.q.Len()
+			// don't copy data from queue unless exporter can accept more, it is very expensive
```
This should be an opt-in `BatchProcessorOption` e.g. `WithDropping` with documentation that this changes the batch processor to drop emitted log records if the queue is full. And document that by default the oldest log records are dropped from the queue.

pellared commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,17 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
-				return ok
-			})
+			qLen := b.q.Len()
+			// don't copy data from queue unless exporter can accept more, it is very expensive
+			if !b.exporter.IsQueueFull() {
+				qLen = b.q.TryDequeue(buf, func(r []Record) bool {
+					ok := b.exporter.EnqueueExport(r)
+					if ok {
+						buf = slices.Clone(buf)
+					}
+					return ok
+				})
+			}
```
Shouldn't we simply add

```go
			if b.exporter.IsQueueFull() {
				continue
			}
```

?

pellared replied:
Why not dropping the records when enqueuing?
Also see https://github.com/open-telemetry/opentelemetry-go/pull/6569/files#r2018253919

pellared commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,17 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
-			})
+			qLen := b.q.Len()
+			// don't copy data from queue unless exporter can accept more, it is very expensive
+			if !b.exporter.IsQueueFull() {
```
There should be a new unit test capturing the changed behavior.

pellared commented on sdk/log/exporter.go:
```diff
@@ -201,6 +200,10 @@ func newBufferExporter(exporter Exporter, size int) *bufferExporter {
 }
 
+func (e *bufferExporter) IsQueueFull() bool {
+	return len(e.input) == cap(e.input)
```
I think this can have a data race as access to `e.input` is not synchronized.
Shouldn't we `IsQueue` full be implemented in`queue` struct? E.g.

```go
func (q *queue) Full() bool {
	q.Lock()
	defer q.Unlock()
	return q.len == q.cap
}
```

pellared reviewed (commented):
I am happy that improving the logs batching processor does not seem to require making changes in the exported API surface :tada:

pellared commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,17 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
-			})
+			qLen := b.q.Len()
+			// don't copy data from queue unless exporter can accept more, it is very expensive
+			if !b.exporter.IsQueueFull() {
+				qLen = b.q.TryDequeue(buf, func(r []Record) bool {
+					ok := b.exporter.EnqueueExport(r)
```
This looks very "dead-locky". I am aware that my comment is not very precise, but I have a feeling that the the logic should reorganized...

Maybe the goroutine responsible for actually exporting the log records should be also responsible for getting/dequeuing the log records from the queue instead of doing it here? This should make sure that we are dequeuing the log records in the nearest place where it is actually needed.

wmdanor replied:
I don't think I have fully grasped race conditions here yet, I need to look into it more.

Do you think adding mutex lock inside `IsQueueFull` can mitigate it for now?
Or it is about the fact that `IsQueueFull` is not really synchronized with `EnqueueExport`?

Not sure I understand reorganising piece. Is it something like passing the queue pointer to buffer so it will internally check if it should try to dequeue? Or more like big rewrite that will remove buffer exporter and move its logic to batch exporter for example?

pellared replied:
> Do you think adding mutex lock inside IsQueueFull can mitigate it for now?

I do not think it will change anything. 

> Or it is about the fact that IsQueueFull is not really synchronized with EnqueueExport?

This and if I recall correctly ForceFlush and Shutdown (writing from mobile and from memory).

> Or more like big rewrite that will remove buffer exporter and move its logic to batch exporter for example?

Rather this.

wmdanor replied:
Ok, I will am going to explore this in a separate PR.

wmdanor replied:
Just small update, I am working on this, there is progress and it may be a little more performant than current solution (with buffer exporter) too, but can't say when this is going to be ready for review as I don't have much of free time now.

dmathieu replied:
Resolving this comment, as this remaining issue will be tackled in a separate PR.

pellared commented:
@wmdanor, do you think it is ready for review?

pellared commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,17 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
-			})
+			qLen := b.q.Len()
+			// Don't copy data from queue unless exporter can accept more, it is very expensive.
+			if !b.exporter.IsQueueFull() {
```
Shouldn't we call `if b.exporter.IsQueueFull()  { continue  }` instead?
Woulnd't we otherwise call `b.pollTrigger <- struct{}{}:` and loop over and over until the queue is not full?

wmdanor replied:
I think we can do that.

At first I was concerned if we are risking to wait potentially for new whole ticker cycle (1 second by default), and 1 second can be both enough time for exporter to be able to accept new data and for processor to drop lots of new records under high load. But then I saw that OnEmit sends signal to pollTrigger if queue.len is bigger than batch size, so it should be ok.

May be I am still missing something. I pushed what you proposed, but I can revert it back if we want to.

pellared replied:
@MrAlias, thoughts?

pellared replied:
I think it may need to be reverted (see the test failures).

wmdanor replied:
Weird that tests did not fail when I ran make precommit before pushing.
I have reverted it now.

wmdanor commented:
> @wmdanor, do you think it is ready for review?

It have noticed that CI failed this morning, so I put it back to draft as I did not have enough time to fix it before going to work.

I saw that you have helped me with resolving those issues, thank you for that.

I have addressed your comment below, so it should be ready now.

Other than that, there is race condition comment you left, should I be trying to add some mutex locks for additional security for this in this PR or do we merge this one as good enough?

pellared commented:
I think this is good enough for a review

pellared commented:
I think after this is addressed: https://github.com/open-telemetry/opentelemetry-go/pull/6569#discussion_r2019126045 it can be reviewed by others.

wmdanor commented:
> I think after this is addressed: [#6569 (comment)](https://github.com/open-telemetry/opentelemetry-go/pull/6569#discussion_r2019126045) it can be reviewed by others.

@pellared pushed the revert.

pellared commented on sdk/log/exporter.go:
```diff
@@ -201,6 +200,10 @@ func newBufferExporter(exporter Exporter, size int) *bufferExporter {
 	}
 }
 
+func (e *bufferExporter) IsQueueFull() bool {
```
What do you think of the following rename?

```suggestion
func (e *bufferExporter) IsInputFull() bool {
```
Otherwise one may think that it has to do something with `type queue`.

pellared replied:
We could also revert the logic and rename to e.g. `IsReady`

MrAlias commented on sdk/log/batch.go:
```diff
@@ -156,13 +156,18 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
+					}
+					return ok
+				})
+			}
```
```suggestion
} else {
   qLen = b.q.Len()
}
